src/journal-remote/header-util.c

```diff
@@ -0,0 +1,170 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "header-util.h"
+#include "strv.h"
+
+/* HTTP header name can contains:
+- Alphanumeric characters: a-z, A-Z, and 0-9
+- The following special characters: - and _ */
+#define VALID_HEADER_NAME_CHARS \
+        DIGITS LETTERS "_"      \
+                       "-"
+
+#define VALID_HEADER_NAME_LENGTH 40
+
+#define VALID_HEADER_VALUE_CHARS \
+        DIGITS LETTERS "_"       \
+                       " "       \
+                       ":"       \
+                       ";"       \
+                       "."       \
+                       ","       \
+                       "\\"      \
+                       "/"       \
+                       "'"       \
+                       "\""      \
+                       "?"       \
+                       "!"       \
+                       "("       \
+                       ")"       \
+                       "{"       \
+                       "}"       \
+                       "["       \
+                       "]"       \
+                       "@"       \
+                       "<"       \
+                       ">"       \
+                       "="       \
+                       "-"       \
+                       "+"       \
+                       "*"       \
+                       "#"       \
+                       "$"       \
+                       "&"       \
+                       "`"       \
+                       "|"       \
+                       "~"       \
+                       "^"       \
+                       "%"
+
+static bool header_name_is_valid(const char *e, size_t n) {
+        if (!e)
```

Comment:
this check should be n == 0 instead,

and please add `assert(e)` afterwards.

because if n is > 0 and e is NULL that's a programming error, and we should assert() on those